stablehlo/conversions/tosa/transforms/StablehloQuantLegalizeToTosaRescale.cpp

```diff
@@ -0,0 +1,257 @@
+      rewriter.getDenseI32ArrayAttr({multiplier}),
+      rewriter.getDenseI8ArrayAttr({static_cast<int8_t>(shift)}),
+      rewriter.getBoolAttr(scale32), rewriter.getBoolAttr(double_round),
+      rewriter.getBoolAttr(perChannel));
+
+  return rescale_op.getResult();
+}
+
+// Creates TOSA rescale op with int32 output
+Value buildRescaleToInt32(PatternRewriter& rewriter, Location loc,
+                          Value inputVal, double inputScale, int64_t inputZp) {
+  auto inputType = dyn_cast<mlir::ShapedType>(inputVal.getType());
+  if (!inputType) {
+    return nullptr;
+  }
+  auto outputType = inputType.clone(rewriter.getI32Type());
+
+  const int32_t scaleWidth = 32;
+  int32_t multiplier, shift;
+  computeMultiplierAndShift(inputScale, multiplier, shift, scaleWidth);
+
+  return buildRescale(rewriter, loc, outputType, inputVal, multiplier, shift,
+                      inputZp,
+                      /* output_zp = */ 0,
+                      /* double_round = */ false,
+                      /* scale32 = */ true,
+                      /* per_channel = */ false);
+}
+
+// Creates TOSA rescale op with int32 input
+Value buildRescaleFromInt32(PatternRewriter& rewriter, Location loc,
+                            mlir::ShapedType outputType, Value inputVal,
+                            double outputScale, int64_t outputZp) {
+  // Input should be int32 type
+  auto inputType = dyn_cast<mlir::ShapedType>(inputVal.getType());
+  if (!inputType || !inputType.getElementType().isInteger(32)) {
+    // expected rescale input element type to be i32
+    return nullptr;
+  }
+
+  const int32_t scaleWidth = 32;
+  int32_t multiplier, shift;
+  computeMultiplierAndShift(outputScale, multiplier, shift, scaleWidth);
+
+  return buildRescale(rewriter, loc, outputType, inputVal, multiplier, shift,
+                      /* input_zp = */ 0, outputZp,
+                      /* double_round = */ false,
+                      /* scale32 = */ true,
+                      /* per_channel = */ false);
+}
+
+template <typename StablehloOp>
+LogicalResult matchAndRewriteAddSub(StablehloOp op, PatternRewriter& rewriter) {
+  Value lhs = op.getLhs();
+  Value rhs = op.getRhs();
+  Value result = op.getResult();
+
+  auto lhsType = dyn_cast<mlir::ShapedType>(lhs.getType());
+  auto rhsType = dyn_cast<mlir::ShapedType>(rhs.getType());
+  auto resultType = dyn_cast<mlir::ShapedType>(result.getType());
+
+  if (!lhsType || !rhsType || !resultType) {
+    return rewriter.notifyMatchFailure(
+        op, "inputs and result tensors should all have shaped tensor types");
+  }
+
+  auto lhsQType =
+      dyn_cast<mlir::quant::UniformQuantizedType>(lhsType.getElementType());
+  auto rhsQType =
+      dyn_cast<mlir::quant::UniformQuantizedType>(rhsType.getElementType());
+  auto resultQType =
+      dyn_cast<mlir::quant::UniformQuantizedType>(resultType.getElementType());
+
+  if (!lhsQType || !rhsQType || !resultQType) {
```

Comment:
```c++
if(!lhsQType && !rhsQType && !resultQType) {
  return rewriter.notifyMatchFailure(op,
                                         "The conversion supports operands/results with per-tensor quantized types only");
}
```

With that we can merge the two returns.

Reply:
I think this feedback is still pending to be applied. 
The only case of interest to us is `when all operands/results are quantized`. If that is not true we can return once with failure. IOW replace
```
  if (!lhsQType || !rhsQType || !resultQType) {
 if (lhsQType || rhsQType || resultQType) {
      return rewriter.notifyMatchFailure(op,
                                         "inputs and result tensors should be "
                                         "all quantized or all un-quantized");
    }
    // not quantized - skip
    return failure();
  }
}
``` 

with

```
if(!lhsQType || !rhsQType || !resultQType) { return rewriter.notifyMatchFailure(op, "The conversion supports operands/results with per-tensor quantized types only"); }
```

Reply:
done